### PR TITLE
[Doc] Refactor runner tutorial

### PR DESCRIPTION
**Pull request by C1rN09**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

New modifications since PR #540 :

- Add a graph to illustrate core idea of `Registry`
- Split and move "Runner Best Practice" into a new chapter, which focuses on config file
- Add explanation & pseudo codes for data flow between `dataloader` `model` and `evaluator`
- Some minor fixes

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: docs/zh_cn/tutorials/runner.md [hunk 1]
```diff
@@ -1,260 +1,527 @@
...
+    ...
+)
 ```
 
-## 通过配置文件使用执行器
+类似上述例子，执行器中的参数大多同时支持两种输入类型。以上两种写法基本是等价的，区别在于：前者以 `dict` 作为输入时，该模块会在**需要时在执行器内部**被构建；而后者是构建完成后传递给执行器。如果你对于注册机制并不了解，下面的示意图展示了它的核心思想：注册器维护着**模块的构建方式**和它的**名字**之间的映射。如果你在使用中发现问题，或者想要进一步了解完整用法，我们推荐阅读[注册器（Registry）](../advanced_tutorials/registry.md)文档。
 
-OpenMMLab 的开源项目普遍使用注册器 + 配置文件的方式来管理和构建模块，MMEngine 中的执行器也推荐使用配置文件进行构建。
-下面是一个通过配置文件使用执行器的例子：
+![Runner Registry 示意图](https://user-images.githubusercontent.com/112053249/199191651-44174d17-0fc5-4443-8d15-76f561ec0585.png)
```
Comment (HAOCHENYE) [inline docs/zh_cn/tutorials/runner.md:256]: 可以补充一下：

看到这你可能仍然很疑惑，为什么我要传入字典让 Runner 来构建实例，这样又有什么好处？如果你有产生这样的疑问，那我们就会很自豪的回答：“没有好处”。事实上直接传入字典的写法也并非最佳实践，而是引出最佳实践的楔子。我们希望你能够通过这一章理解 Runner 是如何基于字典构造的，方便理解我们推荐的——最佳实践。

File: docs/zh_cn/tutorials/runner.md [hunk 2]
```diff
@@ -1,260 +1,527 @@
...
 
-与手动构建模块来使用执行器不同的是，通过调用 Runner 类的 `build_from_cfg` 方法，执行器能够自动读取配置文件中的模块配置，从相应的注册器中构建所需要的模块，用户不再需要考虑训练和测试分别依赖哪些模块，也不需要为了切换训练的模型和数据而大量改动代码。
+OpenMMLab 下游库广泛采用了配置文件的方式。我们将在下个章节，基于上述示例稍微变换，从而展示配置文件——MMEngine 中执行器的最佳实践——的用法。
 
-下面是一个典型的使用配置文件调用 MMClassification 中的模块训练分类器的简单例子：
+</details>
+
+## 执行器最佳实践——配置文件
+
+MMEngine 提供了一套 `Python` 语法的、功能强大的配置文件系统。你可以从之前的示例代码中**近乎**（我们将在下面说明）无缝地转换到配置文件。下面给出一段示例代码：
```
Comment (HAOCHENYE) [inline docs/zh_cn/tutorials/runner.md:278]:
```suggestion
MMEngine 提供了一套支持 `Python` 语法的、功能强大的配置文件系统。你可以从之前的示例代码中**近乎**（我们将在下面说明）无缝地转换到配置文件。下面给出一段示例代码：
```

File: docs/zh_cn/tutorials/runner.md [hunk 3]
```diff
@@ -1,260 +1,527 @@
...
-runner.train()
+你可以在对应模块的教程中找到丰富的说明和示例，你也可以在 [API 文档](mmengine.runner.Runner) 中找到 `Runner` 的所有参数。如果上述两种方式都无法解决你的疑问，你随时可以在我们的 [讨论版](https://github.com/open-mmlab/mmengine/discussions) 中发起话题，帮助我们更好地改进文档。
 
-# 执行测试
-runner.test()
-```
+</details>
+
+<details>
+<summary>我来自 mmdet/mmcls...下游库，为什么例子写法与我接触的不同？</summary>
```
Comment (HAOCHENYE) [inline docs/zh_cn/tutorials/runner.md:270]: MMDet, MMCls。和 MMEngine 风格统一

File: docs/zh_cn/tutorials/runner.md [hunk 4]
```diff
@@ -1,260 +1,527 @@
...
+
+## 基本数据流
+
+```{hint}
+在本章节中，我们将会介绍执行器内部各模块之间的数据传递流向与格式约定。如果你还没有基于 MMEngine 构建一个训练流程，本章节的部分内容可能会比较抽象、枯燥；你也可以暂时跳过，并在将来有需要时结合实践进行阅读。
+```
+
+接下来，我们将**稍微**深入执行器的内部，结合图示来理清其中数据的流向与格式约定。
+
+![基本数据流](https://user-images.githubusercontent.com/112053249/199228350-5f80699e-7fd2-4b4c-ac32-0b16b1922c2e.png)
```
Comment (HAOCHENYE) [inline docs/zh_cn/tutorials/runner.md:376]: 感觉前面可以提一下，如果已经对配置文件很熟悉了，可以直接跳到这一节？

File: docs/zh_cn/tutorials/runner.md [hunk 5]
```diff
@@ -1,260 +1,527 @@
...
+
+上图是执行器的**基本**数据流，其中虚线边框、灰色填充的不同形状代表不同的数据格式，实线方框代表模块或方法。由于 MMEngine 强大的灵活性与可扩展性，你总可以继承某些关键基类并重载其中的方法，因此上图并不总是成立。只有当你没有自定义 `Runner` 或 `TrainLoop` ，并且你的自定义模型没有重载 `train_step`、`val_step` 与 `test_step` 方法时上图才会成立（而这在检测、分割等任务上是常见的，参考[模型](./model.md)教程）。
+
+<details>
+<summary>可以确切地说明每个数据元素的具体类型吗？</summary>
```
Comment (HAOCHENYE) [inline docs/zh_cn/tutorials/runner.md:381]: 数据元素的引入有点突兀？前面好像没提到

File: docs/zh_cn/tutorials/runner.md [hunk 6]
```diff
@@ -1,260 +1,527 @@
 # 执行器（Runner）
 
-深度学习算法的训练、验证和测试通常都拥有相似的流程，因此 MMEngine 提供了执行器以帮助用户简化这些任务的实现流程。 用户只需要准备好模型训练、验证、测试所需要的模块构建执行器，便能够通过简单调用执行器的接口来完成这些任务。用户如果需要使用这几项功能中的某一项，只需要准备好对应功能所依赖的模块即可。
+- [执行器（Runner）](#执行器runner)
```
[the remark this answers is not in the document]
Comment (C1rN09) [reply docs/zh_cn/tutorials/runner.md:3]: 这个是一个自动生成的 table of contents，我感觉还是保留吧

File: docs/zh_cn/tutorials/runner.md (hunk 4 above)
Comment (C1rN09) [reply docs/zh_cn/tutorials/runner.md:376]: 这个我感觉不太需要，如果是已经熟悉的内容，读者应该会自觉扫过/跳过的。加 hint 可能主要是对一些容易让人厌倦的、困惑的、非必要的内容？

File: docs/zh_cn/tutorials/runner.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (mzr1996) [reply docs/zh_cn/tutorials/runner.md:3]: 个人建议不用保留，我们的 docs 模板本身就有右侧边栏的导航了，没必要再搞一个 toc

[the remark this answers is not in the document]
Comment (C1rN09) [reply docs/zh_cn/tutorials/runner.md]: 有道理

**Review by HAOCHENYE — APPROVED** [review]

File: docs/zh_cn/tutorials/runner.md [hunk 7]
```diff
@@ -1,260 +1,523 @@
...
-test_evaluator = Evaluator(metric)
-
-# 初始化执行器
-runner = Runner(model=model, test_dataloader=test_dataloader, test_evaluator=test_evaluator,
-                load_from='./train_mnist/epoch_3.pth', work_dir='./test_mnist')
-
-# 执行测试
-runner.test()
+<details>
+<summary>为什么有些传入参数是`dict`？</summary>
```
Comment (zhouzaida) [inline docs/zh_cn/tutorials/runner.md:216]:
```suggestion
<summary>为什么有些传入参数是 dict？</summary>
```

File: docs/zh_cn/tutorials/runner.md [hunk 8]
```diff
@@ -1,260 +1,523 @@
...
+
+是的，这与 MMEngine 的风格相关。在 MMEngine 中我们提供了两种不同风格的执行器构建方式：a）基于手动构建的，以及 b）基于注册机制的。如果你感到迷惑，下面的例子将给出一个对比：
+
+```python
+from mmengine.model import BaseModel
+from mmengine.runner import Runner
+from mmengine.registry import MODELS # 模型根注册器，你的自定义模型需要注册到这个根注册器中
```
Comment (zhouzaida) [inline docs/zh_cn/tutorials/runner.md:223]:
```suggestion
from mmengine.registry import MODELS # 模型根注册器，你的自定义模型需要注册到这个根注册器中

```

File: docs/zh_cn/tutorials/runner.md [hunk 9]
```diff
@@ -1,260 +1,524 @@
...
+
+<details>
+<summary>我可以看出红线是训练流程，蓝线是验证/测试流程，但绿线是什么？</summary>
+
+在目前的执行器流程中，`'tensor'` 模式的输出并未被使用，大多数情况下用户无需实现。但一些情况下输出中间结果可以方便地进行 debug
+
+</details>
+
+<details>
+<summary>如果我重载了`train_step`等方法，上图会完全失效吗？</summary>
```
Comment (zhouzaida) [inline docs/zh_cn/tutorials/runner.md:446]:
```suggestion
<summary>如果我重载了 train_step 等方法，上图会完全失效吗？</summary>
```

**Review by zhouzaida — APPROVED** [review]

**Review by RangiLyu — APPROVED** [review]

**Review by tonysy — APPROVED** [review]
LGTM

**Pull request merged**